compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,857 @@
+}
+
+void MLIContext::emitClientWrapper(FnSymbol* fn) {
+  std::string gen;
+
+  this->setOutput(&this->fiClientBundle);
+  fn->codegenHeaderC();
+
+  gen += scope_begin;
+  gen += this->genClientsideRPC(fn);
+  gen += scope_end;
+  gen += "\n";
+
+  this->write(gen);
+  
+  return;
+}
+
+void MLIContext::emitServerWrapper(FnSymbol* fn) {
+  std::string gen;
+
+  // Big long, silly block of manual code generation.
+  gen += this->genComment(toString(fn));
+  gen += "int64_t chpl_mli_swrapper_";
+  gen += this->genFuncNumericID(fn);
+  gen += "(void)";
+  gen += scope_begin;
+
+  gen += this->genServersideRPC(fn);
+  gen += "return 0;\n";
+ 
+  gen += scope_end;
+  gen += "\n";
+
+  this->setOutputAndWrite(&this->fiServerBundle, gen);
+ 
+  return;
+}
+
+std::string MLIContext::genDebugPrintCall(FnSymbol* fn, const char* pfx) {
+  std::string msg;
+
+  msg += "Calling: ";
+  msg += toString(fn);
+
+  return this->genDebugPrintCall(msg.c_str(), pfx);
+}
+
+std::string MLIContext::genDebugPrintCall(const char* msg, const char* pfx) {
+  std::string gen;
+
+  gen += "printf(\"";
+
+  if (pfx && strcmp(pfx, "")) {
+    gen += pfx;
+    gen += " ";
+  }
+
+  gen += "%s\\n\", \"";
+  gen += msg;
+  gen += "\");\n";
+
+  return gen;
+}
+
+std::string MLIContext::genFuncNumericID(FnSymbol* fn) {
+  return str((int64_t) fn->id);
+}
+
+std::string MLIContext::genServerWrapperCall(FnSymbol* fn) {
+  std::string gen;
+
+  gen += "chpl_mli_swrapper_";
+  gen += this->genFuncNumericID(fn);
+  gen += "();\n";
+
+  return gen;
+}
+  
+std::string
+MLIContext::genServerDispatchSwitch(const std::vector<FnSymbol*>& fns) {
+  std::string gen;
+
+  gen += "int64_t chpl_mli_sdispatch";
+  gen += "(int64_t function)";
+  gen += scope_begin;
+  gen += this->genNewDecl("int", "err");
+  gen += "switch (function)";
+  gen += scope_begin;
+
+  for_vector(FnSymbol, fn, fns) {
+    INT_ASSERT(this->shouldEmit(fn));
+
+    gen += "case ";
+    gen += this->genFuncNumericID(fn);
+    gen += ": ";
+
+    gen += scope_begin;
+    
+    if (this->debugPrintlines) {
+      gen += this->genDebugPrintCall(fn, "[Server]");
+    }
+    
+    gen += "err = ";
+    gen += this->genServerWrapperCall(fn);
+    gen += scope_end;
+    gen += "break;\n";
+  }
+
+  gen += "default: return CHPL_MLI_ERROR_NOFUNC; break;\n";
+  gen += scope_end;
+  gen += "return err;\n";
+  gen += scope_end;
+
+  return gen;
+}
+
+//
+// This filter will change as we support more and more type classes.
+//
+bool MLIContext::isSupportedType(Type* t) {
+  return (isPrimitiveScalar(t));
+}
+
+void MLIContext::verifyPrototype(FnSymbol* fn) {
+
+  if (fn->retType != dtVoid && not isSupportedType(fn->retType)) {
+    USR_FATAL("MLI does not support code generation for type", fn->retType);
+  }
+
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+    if (not this->isSupportedType(as->type)) {
+      USR_FATAL("MLI does not support code generation for type", as->type);
```

Comment:
Ditto these two (I won't explicitly call out additional ones)

Reply:
Hopefully I've addressed all of these by now!